lib/hallux/internal/finger_tree.ex

```diff
@@ -212,23 +212,26 @@ defmodule Hallux.Internal.FingerTree do
   def take_until(t, p), do: split(t, p) |> elem(0)
   def drop_until(t, p), do: split(t, p) |> elem(1)
 
+  # For the view functions, it is often necessary to evaluate
+  # the :m field one level to get the cached monoid value in there.
+
   def view_l(%Empty{}), do: nil
   def view_l(%Single{monoid: mo, x: x}), do: {x, %Empty{monoid: mo}}
 
   def view_l(%Deep{l: %One{a: x}, m: m, r: sf}),
-    do: {x, rot_l(m, sf)}
+    do: {x, rot_l(m.(), sf)}
```

Comment:
Since `m` is being forced here, this is still as strict as before, right? 

I guess that to achieve the O(1) in `head`/`tail` (when they are implemented) one would need to indeed return the thunk unevaluated, so that those functions would not force it, and one would need to force it to get a `Tree` back.

It looks like it'd have to be something like this for `head`/`tail`:

```elixir
def view_l(%Deep{l: %One{a: x}, m: m, r: sf}),
    do: {x, fn -> rot_l(m, sf) end}
```

and force `m` inside each function as needed.

Reply:
Makes sense, that simplifies the whole story a lot more.